client/docs/plugin_model.md

```diff
@@ -0,0 +1,169 @@
+`State` as a parameter to the `invoke` method.  The important
+difference between the different operations is that each type takes
+unique additional parameters. The `invoke` method in the
+`contract_op_base` takes an additional `Session` object with
+information about the enclave service connection that should be
+used. The `contract_command_base` uses a `Context` object to represent
+the configuration of a contract object and its relationship to other
+contract objects. Scripts, take no additional parameters beyond the
+`State`.
+
+Functionality defined in the pluging can be mapped to a `pdo-shell`
+command by defining an entry point and binding it to a command
+attribute in the shell class. Each type of operation (contract
+operations, contract commands and scripts) has a class-specific
+generator that creates a function for the shell,
+`create_shell_command`. The generated commands can be mapped into the
+`pdo-shell` using `bind_shell_command`.
+
+## Base Classes ##
+
+### State ###
+
+Implements the client and contract configuration as a multi-tiered
+dictionary with a few special twists.
+
+State is, in a sense, the root of all other classes. The current
+context and set of bindings are stored in state.
+
+The primary methods for state allow for direct lookup using a list of
+keys (`get`) and an "intelligent" lookup that examines the entries to
+look for special values (`expand`). There are two forms of special
+values in state: links and substitutions. A link (a state value that
+begins with '@') expands to the value of the referenced state entry
+including all of the structure. A substituion (begins with '$')
+expands to a string value.
+
+### Bindings ###
+
+Implements a simple variable store for `pdo-shell` and Python
+scripts. A value may be bound to a variable and a variable may be
+expanded into the corresponding value. This is generally done through
+the `${var}` syntax in the `pdo-shell` though it also happens during
+the loading of configuration files.
+
+When the `pdo-shell` invokes an operation through a script invocation,
+a new binding is created through the `Clone` operation. All variables
+*except* those that start with `_` will be copied into the new
+Binding. That is, you may consider variables that begin with `_` as
+variables local to a particular script.
+
+Upon return from a script invocation, the value of all variables will
+be copied into the calling environment except those variables that
+begin with `_`.
+
+Bindings are most often used to pass settings into a script, to
+communicate the results of a script to the calling context, and for
+saving temporary values that can be passed to other operations.
+
+The initial set of bindings is created from the `Bindings` value in
+state (which generally comes from the `Bindings` field in the
+configuration file.
+
+### SessionParameters ###
+
+Captures the parameters for interacting with a specific contract. This
```

Comment:
Is this really referring to a specific _instance_ of a contract?

Reply:
yes